drivers: display: uc81xx: add support for uc8175

xqinx opened the pull request:
Add support for uc8175 display driver. uc8175 has a slightly different command/data length requirements for certain registers, namely TRES and PTL, compared to uc8176/uc8179

This commit refactors the driver code and such that setting TRES and PTL registers are now done by function pointers provided by config->quirks, by the same token as how it is done for the CDI register

jfischer-no commented on drivers/display/uc81xx.c:
```diff
@@ -25,7 +25,7 @@ LOG_MODULE_REGISTER(uc81xx, CONFIG_DISPLAY_LOG_LEVEL);
  */
 
-#define UC81XX_PIXELS_PER_BYTE		8U
+#define UC81XX_PIXELS_PER_BYTE 8U
```
No.

jfischer-no commented on drivers/display/uc81xx.c:
```diff
@@ -102,8 +104,8 @@ static inline void uc81xx_busy_wait(const struct device *dev)
 
-static inline int uc81xx_write_cmd(const struct device *dev, uint8_t cmd,
-				   const uint8_t *data, size_t len)
+static inline int uc81xx_write_cmd(const struct device *dev, uint8_t cmd, const uint8_t *data,
+				   size_t len)
```
No.

jfischer-no commented on drivers/display/uc81xx.c:
```diff
@@ -142,9 +144,8 @@ static inline int uc81xx_write_cmd(const struct device *dev, uint8_t cmd,
-static inline int uc81xx_write_cmd_pattern(const struct device *dev,
-					   uint8_t cmd,
-					   uint8_t pattern, size_t len)
+static inline int uc81xx_write_cmd_pattern(const struct device *dev, uint8_t cmd, uint8_t pattern,
+					   size_t len)
```
No, to all reformatting.

jfischer-no reviewed (changes requested):
Please drop all of the necessary code reformatting,  then we can continue with the review.

xqinx committed:
drivers: display: uc81xx: add support for uc8175

Add support for uc8175 display driver. uc8175 has a slightly
different command/data length requirements for certain registers,
namely TRES and PTL, compared to uc8176/uc8179

This commit refactors the driver code and such that setting TRES and PTL
registers are now done by function pointers provided by config->quirks,
by the same token as how it is done for setting CDI register

Signed-off-by: Xiao Qin <xiaoq@google.com>

xqinx commented:
> Please drop all of the necessary code reformatting, then we can continue with the review.

Got it, thank you!

jfischer-no commented on drivers/display/uc81xx.c:
```diff
@@ -654,6 +631,122 @@ static int uc81xx_init(const struct device *dev)
+#if DT_HAS_COMPAT_STATUS_OKAY(ultrachip_uc8175)
+static int uc8175_set_cdi(const struct device *dev, bool border)
+{
+	const struct uc81xx_config *config = dev->config;
+	const struct uc81xx_data *data = dev->data;
+	const struct uc81xx_profile *p = config->profiles[data->profile];
+	uint8_t cdi = UC8175_CDI_VBD1 | UC8175_CDI_DDX0 | (p ? (p->cdi & UC8175_CDI_CDI_MASK) : 0);
+
+	if (!p || !p->override_cdi) {
+		return 0;
+	}
+
+	if (!border) {
+		/* Floating border */
+		cdi |= UC8175_CDI_VBD1 | UC8175_CDI_VBD0;
+	}
+
+	LOG_DBG("CDI: %#hhx", cdi);
+	return uc81xx_write_cmd_uint8(dev, UC81XX_CMD_CDI, cdi);
+}
```
If the differences are too great and uc8179_set_cdi() cannot be used, it should at least be implemented in the same way as uc8179_set_cdi().

xqinx replied:
yep, I just check against the datasheet and it's completely fine to use uc8176_set_cdi for uc8175, the extra bit in the CDI mask will be silently ignored by the chip, all good.

jfischer-no commented on drivers/display/uc81xx.c:
```diff
@@ -654,6 +631,122 @@ static int uc81xx_init(const struct device *dev)
+				   const struct display_buffer_descriptor *desc)
+{
+	const struct uc81xx_config *config = dev->config;
+
```
Please remove this empty line.

xqinx replied:
done

jfischer-no commented on drivers/display/uc81xx_regs.h:
```diff
@@ -106,26 +113,8 @@
 #define UC8179_CDI_DDX1				BIT(1)
 #define UC8179_CDI_DDX0				BIT(0)
 
-struct uc81xx_tres {
-	uint16_t hres;
-	uint16_t vres;
-} __packed;
-
-BUILD_ASSERT(sizeof(struct uc81xx_tres) == 4);
-
-struct uc81xx_ptl {
-	uint16_t hrst;
-	uint16_t hred;
-	uint16_t vrst;
-	uint16_t vred;
-	uint8_t flags;
-} __packed;
-
-BUILD_ASSERT(sizeof(struct uc81xx_ptl) == 9);
-
```
Instead, please rename it to uc81xx_tres16, uc81xx_ptl16 and add uc81xx_tres8 and uc81xx_ptl8.

xqinx replied:
done

andysan reviewed (changes requested):
Thanks for this change. This looks like a good way of implementing UC8175 support. I have a minor change request (see comment below), but it looks good other than that.

xqinx commented on drivers/display/uc81xx.c:
```diff
@@ -425,12 +413,6 @@ static int uc81xx_write(const struct device *dev, const uint16_t x, const uint16
 	__ASSERT(!(desc->width % UC81XX_PIXELS_PER_BYTE),
 		 "Buffer width not multiple of %d", UC81XX_PIXELS_PER_BYTE);
 
-	if ((y_end_idx > (config->height - 1)) ||
```
Sounds good, thank you!

andysan commented on drivers/display/uc81xx.c:
```diff
@@ -403,13 +401,7 @@ static int uc81xx_write(const struct device *dev, const uint16_t x, const uint16
-		.vred = sys_cpu_to_be16(y_end_idx),
-		.flags = UC81XX_PTL_FLAG_PT_SCAN,
-	};
+
```
This newline seems redundant.

andysan commented on drivers/display/uc81xx.c:
```diff
@@ -425,8 +417,7 @@ static int uc81xx_write(const struct device *dev, const uint16_t x, const uint16
 
-	if ((y_end_idx > (config->height - 1)) ||
-	    (x_end_idx > (config->width - 1))) {
+	if ((y_end_idx > (config->height - 1)) || (x_end_idx > (config->width - 1))) {
```
It'd be nice if you could keep the same formatting as as before to keep the diff as small as possible. That keeps the git annotate log clean and avoids merge conflicts in the future.

xqinx replied:
ah sorry about that, fixed :)

andysan reviewed (commented):
Thanks for the updated change. I'm OK with this version of the change, but I would prefer it if you could address the two comments I have added in this review.

andysan reviewed (approved):
Thanks for the quick turnaround. This looks good to me.

jfischer-no reviewed (approved):
Thanks

carlescufi commented:
@andysan Thanks! please consider becoming a collaborator in this area:
https://docs.zephyrproject.org/latest/project/project_roles.html#collaborator
